### PR TITLE
feat: add DPoP sender-constrained token support (RFC 9449)

### PR DESCRIPTION
**Pull request by davidbemer**
## Summary

- Adds `src/SDK/Token/DPoP.php` implementing RFC 9449 DPoP proof validation (signature verification for RSA, EC P-256/384/521, RSA-PSS, and EdDSA; JWK thumbprint check; htm/htu/iat/ath claim validation)
- Adds `DescopeSDK::validateDPoP()` public method — a thin wrapper for server-side use after `verify()`; no-op when the session token has no `cnf.jkt` claim
- Adds `src/tests/DPoPTest.php` with PHPUnit tests covering valid RSA and EC proofs, all structural error paths, and URL normalisation edge cases

Mirrors the Go SDK implementation: https://github.com/descope/go-sdk/pull/737

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/SDK/Token/DPoP.php [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+    private static function base64urlDecodeJson(string $data): array
+    {
+        $decoded = self::base64urlDecode($data);
+        $result = json_decode($decoded, true);
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new TokenException('Invalid JSON in JWT part: ' . json_last_error_msg());
+        }
+        if (!is_array($result)) {
+            throw new TokenException('JWT part JSON did not decode to an array');
+        }
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:209]: Fixed. Added `if (!is_array($result)) { throw new TokenException('JWT part JSON did not decode to an array'); }` after the `json_last_error()` check. The function now guarantees it only returns a true PHP array.

File: src/SDK/Token/DPoP.php [hunk 2]
```diff
@@ -0,0 +1,702 @@
...
+        if ($alg === 'EdDSA' && $kty !== 'OKP') {
+            throw new TokenException('alg EdDSA requires kty=OKP, got kty=' . $kty);
+        }
+
+        // Step 14-15: Import JWK and verify signature
+        $publicKey = self::importJwkAsPublicKey($jwk, $alg);
+        $signingInput = $parts[0] . '.' . $parts[1];
+        $signature = self::base64urlDecode($parts[2]);
+        self::verifyDpopSignature($signingInput, $signature, $publicKey, $jwk, $alg);
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:126]: Fixed. Added an explicit alg/kty cross-check block before `importJwkAsPublicKey`: RS*/PS* algs require `kty=RSA`, ES* require `kty=EC`, and EdDSA requires `kty=OKP`. A `TokenException` is thrown on mismatch.

File: src/SDK/Token/DPoP.php [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+
+        // Step 16: Decode payload
+        $payload = self::base64urlDecodeJson($parts[1]);
+
+        // Steps 17-19: Validate required claims
+        if (empty($payload['jti'])) {
+            throw new TokenException('missing jti');
+        }
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:134]: Deferred — jti replay protection (RFC 9449 §11.1) is intentionally out of scope for a stateless SDK. This matches the go-sdk reference implementation. Added a note in the class docblock explaining the omission and directing callers to maintain their own jti store (e.g., a short-lived cache keyed on jti with TTL matching the iat window) if replay protection is required.

File: src/SDK/Token/DPoP.php [hunk 4]
```diff
@@ -0,0 +1,702 @@
...
+        // Step 1-2: Check proof length
+        $dpopProof = trim($dpopProof);
+        if (strlen($dpopProof) > self::MAX_PROOF_LEN) {
+            throw new TokenException('DPoP proof exceeds maximum length');
+        }
+
+        // Step 3: Require proof when token is DPoP-bound
+        if ($dpopProof === '') {
+            throw new TokenException('DPoP proof required: access token is DPoP-bound (cnf.jkt present)');
+        }
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:76]: Fixed. All `throw new \Exception(` calls in `DPoP.php` have been replaced with `throw new TokenException(`, consistent with `Extractor.php` and `Verifier.php`. Added `use Descope\SDK\Exception\TokenException;` at the top of the file.

File: src/SDK/DescopeSDK.php [hunk 5]
```diff
@@ -226,6 +227,27 @@ public function logoutAll(?string $refreshToken = null): void
...
+     *
+     * If the session token does not contain a cnf.jkt claim, this method
+     * returns immediately (the token is not DPoP-bound).
+     *
+     * @param  string $sessionToken The session JWT (already verified).
+     * @param  string $dpopProof    The value of the DPoP HTTP header sent by the client.
+     * @param  string $method       The HTTP method of the incoming request (e.g. "GET", "POST").
+     * @param  string $requestUrl   The full URL of the incoming request.
+     * @throws \Descope\SDK\Exception\TokenException If the DPoP proof is missing or invalid.
+     */
```
Comment (davidbemer) [inline src/SDK/DescopeSDK.php:241]: Fixed. Updated `@throws` docblock to `\Descope\SDK\Exception\TokenException` to match what `DPoP::validateProof()` now throws.

**Review by davidbemer — COMMENTED** [review]
Inline replies to all review comments (details in the PR-level comment above).

File: src/tests/DPoPTest.php [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+use PHPUnit\Framework\TestCase;
+use Descope\SDK\Token\DPoP;
+
+/**
+ * Unit tests for DPoP (RFC 9449) proof validation.
+ *
+ * These tests exercise the static helper methods directly without
+ * requiring a network connection or live Descope credentials.
+ */
+final class DPoPTest extends TestCase
```
Comment (davidbemer) [inline src/tests/DPoPTest.php:14]: Fixed. Added `<file>src/tests/DPoPTest.php</file>` to the test suite in `phpunit.xml`.

File: README.md [hunk 7]
```diff
@@ -255,6 +255,32 @@ print_r($response);
...
+```php
+// 1. Verify the session token as usual
+$descopeSDK->verify($sessionToken);
+
+// 2. If the token is DPoP-bound, validate the DPoP proof
+//    (does nothing if the token has no cnf.jkt claim)
+$descopeSDK->validateDPoP(
+    $sessionToken,               // the verified session JWT
+    $_SERVER['HTTP_DPOP'] ?? '', // DPoP header sent by the client (empty string if absent)
+    $_SERVER['REQUEST_METHOD'],  // e.g. "GET" or "POST"
```
Comment (davidbemer) [inline README.md:273]: Fixed. Changed `$_SERVER['HTTP_DPOP']` to `$_SERVER['HTTP_DPOP'] ?? ''` so the example handles the case where the DPoP header is absent without a PHP notice.

File: src/SDK/Token/DPoP.php [hunk 8]
```diff
@@ -0,0 +1,702 @@
...
+        $digestLen = strlen($digest);
+
+        // Decrypt the signature using the RSA public key (raw RSA operation)
+        $decrypted = '';
+        $decryptResult = openssl_public_decrypt($signature, $decrypted, $publicKey, OPENSSL_NO_PADDING);
+        if (!$decryptResult) {
+            throw new TokenException('RSA-PSS signature verification failed (decrypt step)');
+        }
+
+        // Verify PSS encoding manually
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:508]: Fixed. Removed the unused `$opensslAlgMap` array and the stale comment describing an `openssl_verify` fallback. The function now only contains the actual working implementation: raw RSA public decrypt + manual EMSA-PSS verification.

**Review by davidbemer — COMMENTED** [review]
Follow-up inline replies for rawEcSigToDer and verifyPss.

File: src/SDK/Token/DPoP.php [hunk 9]
```diff
@@ -0,0 +1,702 @@
...
+
+        if ($kty === 'EC') {
+            // JWTs use raw R||S; OpenSSL needs DER-encoded ECDSA signature
+            $signature = self::rawEcSigToDer($signature);
+        }
+
+        $result = openssl_verify($signingInput, $signature, $publicKey, $opensslAlg);
+        if ($result !== 1) {
+            throw new TokenException('DPoP proof signature verification failed');
+        }
```
Comment (davidbemer) [inline src/SDK/Token/DPoP.php:443]: Fixed. Changed the `$r` guard from `if (ord($r[0]) > 0x7f)` to `if (strlen($r) === 0 || ord($r[0]) > 0x7f)`, matching the identical pattern already applied to `$s`. All-zero input for `$r` will now safely prepend `\x00` rather than triggering a warning on `$r[0]`.

File: src/SDK/DescopeSDK.php [hunk 10]
```diff
@@ -226,6 +227,27 @@ public function logoutAll(?string $refreshToken = null): void
...
+    public function validateDPoP(
+        string $sessionToken,
+        string $dpopProof,
+        string $method,
+        string $requestUrl
+    ): void {
+        DPoP::validateProof($dpopProof, $method, $requestUrl, $sessionToken);
```
Comment (davidbemer) [inline src/SDK/DescopeSDK.php:248]: Fixed. `DPoP::validateProof()` and all other throw sites in `DPoP.php` now throw `TokenException` (imported via `use Descope\SDK\Exception\TokenException`). The `@throws` docblock on `validateDPoP()` has been updated accordingly.

File: README.md (hunk 7 above)
Comment (davidbemer) [inline README.md:273]: Fixed. Changed to `$_SERVER['HTTP_DPOP'] ?? ''` so the example handles the absent-header case without a PHP notice. Also updated the `validateDPoP` throws description to `\Descope\SDK\Exception\TokenException`.

**Comment by dorsha** [comment]
@davidbemer what is the status of this PR?
